hack/generator/pkg/codegen/pipeline_remove_embedded_resources.go

```diff
@@ -0,0 +1,681 @@
+	}
+
+	return &propertiesTypeName, nil
+}
+
+func extractStatusPropertiesTypeNameOrNil(types astmodel.Types, resource *astmodel.ResourceType) (*astmodel.TypeName, error) {
+	statusName, ok := astmodel.AsTypeName(resource.StatusType())
+	if !ok {
+		return nil, nil
+	}
+
+	return followTypeNameExtractPropertiesType(types, statusName)
+}
+
+func extractSpecPropertiesTypeNameOrNil(types astmodel.Types, resource *astmodel.ResourceType) (*astmodel.TypeName, error) {
+	specName, ok := astmodel.AsTypeName(resource.SpecType())
+	if !ok {
+		return nil, errors.Errorf("resource spec was not a TypeName")
+	}
+	return followTypeNameExtractPropertiesType(types, specName)
+}
+
+type resourceRemovalVisitorContext struct {
+	resource      astmodel.TypeName
+	depth         int
+	modifiedTypes astmodel.Types
+}
+
+func (e resourceRemovalVisitorContext) WithMoreDepth() resourceRemovalVisitorContext {
+	e.depth += 1
+	return e
+}
+
+type embeddedResourceRemover struct {
+	types                    astmodel.Types
+	resourceToSubresourceMap map[astmodel.TypeName]astmodel.TypeNameSet
+	resourcePropertiesTypes  astmodel.TypeNameSet
+	resourceStatusTypes      astmodel.TypeNameSet
+	typeSuffix               string
+	typeFlag                 astmodel.TypeFlag
+}
+
+func makeEmbeddedResourceRemover(types astmodel.Types) (embeddedResourceRemover, error) {
+	resourceStatusTypes := findAllResourceStatusTypes(types)
+	resourceToSubresourceMap, err := findSubResourcePropertiesTypeNames(types)
+	if err != nil {
+		return embeddedResourceRemover{}, errors.Wrap(err, "couldn't find subresource \"Properties\" type names")
+	}
+
+	resourcePropertiesTypes, err := findAllResourcePropertiesTypes(types)
+	if err != nil {
+		return embeddedResourceRemover{}, errors.Wrap(err, "couldn't find resource \"Properties\" type names")
+	}
+
+	remover := embeddedResourceRemover{
+		types:                    types,
+		resourceToSubresourceMap: resourceToSubresourceMap,
+		resourcePropertiesTypes:  resourcePropertiesTypes,
+		resourceStatusTypes:      resourceStatusTypes,
+		typeSuffix:               "SubResourceEmbedded",
+		typeFlag:                 astmodel.TypeFlag("embeddedSubResource"), // TODO: Instead of flag we could just use a map here if we wanted
+	}
+
+	return remover, nil
+}
+
+func (e embeddedResourceRemover) MakeEmbeddedResourceRemovalTypeVisitor() astmodel.TypeVisitor {
+	visitor := astmodel.MakeTypeVisitor()
+	visitor.VisitObjectType = func(this *astmodel.TypeVisitor, it *astmodel.ObjectType, ctx interface{}) (astmodel.Type, error) {
+		typedCtx := ctx.(resourceRemovalVisitorContext)
+
+		if typedCtx.depth <= 2 {
+			// If we are not at sufficient depth, don't bother checking for subresource references. The resource itself and its spec type
+			// will not refer to a subresource. There are some instances of resources (such as Microsoft.Web v20160801 Sites) where the resource
+			// and some child resources reuse the same "Properties" type, which could false some of the logic below
```

Comment:
This whole comment was written in the depths of madness as I was writing the code. I've clarified it significantly based on both of your feedback, to:
```
// Avoid removing top level "Properties", which we never want to do. This is needed because
// there are some resources (such as Microsoft.Web v20160801 Sites) where the resource
// and some child resources reuse the same "Properties" type. This causes 
// the logic below to think that a resource is its own subresource. Without this
// check the entire resource would be removed, leaving nothing.
```